### DISCUSSION TITLE
Welcome to react-graph-vis Discussions!

### DISCUSSION THREAD

**Discussion by crubier** [crubier/react-graph-vis#105] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ZachHaber** [comment]
Hey, just wanted to share what I put together for a project: https://gist.github.com/ZachHaber/feb0432fe395a303a4f10671877e5e70

We ran into the issue https://github.com/crubier/react-graph-vis/issues/92#issuecomment-818279270 which prevented use of strict mode, so I modified the code to using a function component with TS definitions.

It's far too large of a modification to reasonably put in a PR for the change as it's a full rewrite. But I wanted to share it because it fixed the issues with hot reloading we were running into, and fixed some other issues with the graph not updating.

**Reply by Rocamonde** [reply]
Hi @ZachHaber, apparently your Gist does not include options such as getNetwork, which we are currently using. I'll try to add it myself and suggest an edit if I manage to do so. Do let me know if you have any ideas about how to do it.

**Reply by ZachHaber** [reply]
@Rocamonde, I actually set up ForwardRef to give access to the Network. So you should be able to pass in either a Ref or a function to the `ref` prop of `VisGraph` to gain access to the Network.

**Reply by Rocamonde** [reply]
Hi @ZachHaber, that's useful! I did notice, however, that setting it up requires a bit of extra work. Here's the code I used, in case it's useful for anyone:

```typescript
const networkRef = useCallback<
    (network: Network | null | undefined) => void
>((network) => {
    if (network !== undefined && network !== null) {
        window.addEventListener("resize", () => {
            network.setOptions({
                width: window.innerWidth.toString(),
                height: window.innerHeight.toString(),
            });
        });
    }
}, []);
```

Maybe it's worth adding a custom hook that wraps useCallback and does the undefined and null type checks for us.

Then pass `networkRef` as a ref within `<VisGraph/>`.

**Reply by ZachHaber** [reply]
@Rocamonde You should put event listeners in an effect so you can clean them up.

```
const networkRef = useRef<Network | undefined>(undefined);
useEffect(() => {
  const resize = () => {
    if (!networkRef.current) {
      return;
    }
    networkRef.current.setOptions({
      width: window.innerWidth.toString(),
      height: window.innerHeight.toString(),
    });
  };
  window.addEventListener("resize", resize);
  return () => {
    window.removeEventListener("resize", resize);
  };
}, []);

return <VizGraph ref={networkRef} />

```

That being said, one of the ways I used this was I made a parent component which uses a ResizeObserver on a wrapping div to set the size of the network.

**Reply by Rocamonde** [reply]
@ZachHaber when I used a use ref and use effect the ref was always undefined, apparently because the use effect was called before the ref was attached, and because the ref is mutated instead of replaced the hook didn’t get called again, so I switched to using use callback. Any idea why this might be or how I could fix it?

**Reply by ZachHaber** [reply]
@Rocamonde That's just how refs work. That's why I set this example up the way I did. I always attach the event listener, but then in the event listener itself, I check for nullish values before doing the actual resize. Since the ref mtuates instead of re-rendering, the ref will always be the current value in the listener.

**Reply by vtrphan** [reply]
Thanks for the rewrite, i tried and somehow i cannot use interactions with the graph anymore (drag, zoom, select nodes etc. doesn't work ). Any idea why it doesn't work? Is there an issue with the rewrite?

**Reply by ZachHaber** [reply]
@vtrphan I've done all of those things with the re-write. So, we'd probably need more info to figure that out.
